src/Orleans.Core/Networking/Connection.cs

```diff
@@ -0,0 +1,375 @@
+                    this.RemoteEndpoint);
+            }
+
+            this.CloseInternal(exception);
+        }
+
+        /// <summary>
+        /// Called immediately prior to transporting a message.
+        /// </summary>
+        /// <param name="msg"></param>
+        /// <returns>Whether or not to continue transporting the message.</returns>
+        protected abstract bool PrepareMessageForSend(Message msg);
+
+        protected abstract void OnMessageSerializationFailure(Message msg, Exception exc);
+
+        protected abstract void RetryMessage(Message msg, Exception ex = null);
+
+        private void CloseInternal(ConnectionAbortedException exception)
+        {
+            try
+            {
+                if (!this.IsValid) return;
+
+                lock (this.lockObj)
+                {
+                    if (!this.IsValid) return;
+                    this.IsValid = false;
+                }
+
+                // Try to gracefully stop the reader/writer loops.
+                this.Context.Transport.Input.CancelPendingRead();
+                this.Context.Transport.Output.CancelPendingFlush();
+                this.outgoingMessageWriter.TryComplete();
+
+                if (exception == null)
+                {
+                    this.Context.Abort();
+                }
+                else
+                {
+                    this.Context.Abort(exception);
+                }
+            }
+            catch
+            {
+            }
+        }
+
+        public void Send(Message message)
+        {
+            if (!this.outgoingMessageWriter.TryWrite(message))
+            {
+                this.RerouteMessage(message);
+            }
+        }
+
+        protected abstract void OnReceivedMessage(Message message);
+
+        protected abstract void OnReceiveMessageFailure(Message message, Exception exception);
+
+        protected abstract void OnSendMessageFailure(Message message, string error);
+
+        private async Task ProcessIncoming()
+        {
+            PipeReader input = default;
+            var serializer = this.serviceProvider.GetRequiredService<IMessageSerializer>();
+            try
+            {
+                if (this.Log.IsEnabled(LogLevel.Debug))
+                {
+                    this.Log.LogDebug(
+                        "Starting to process messages from remote endpoint {RemoteEndPoint} to local endpoint {LocalEndPoint}",
+                        this.RemoteEndpoint,
+                        this.LocalEndpoint);
+                }
+
+                input = this.Context.Transport.Input;
+                var requiredBytes = 0;
+                Message message = default;
+                while (true)
+                {
+                    var readResultTask = input.ReadAsync();
+                    var readResult = readResultTask.IsCompletedSuccessfully ? readResultTask.GetAwaiter().GetResult() : await readResultTask;
```

Comment:
Did you see a measurable performance difference doing this rather than just awaiting?